### DISCUSSION TITLE
Nameof with generic type definitions

### DISCUSSION THREAD

**Discussion by dylanrafael05** [dotnet/csharplang#5115] [Language Ideas] [answered]
When using the nameof operator with a generic type, type arguments are ignored.

Consider the following program:
```cs
Console.WriteLine(nameof(Generic));                 // prints "Generic"
Console.WriteLine(nameof(Generic<object>));         // prints "Generic"
Console.WriteLine(nameof(Generic<object, object>)); // prints "Generic"

Console.WriteLine(nameof(Generic.field));                 // prints "field"
Console.WriteLine(nameof(Generic<object>.field));         // prints "field"
Console.WriteLine(nameof(Generic<object, object>.field)); // prints "field"

Console.WriteLine(nameof(Generic<int>) == nameof(Generic<string>));    // prints "True"
Console.WriteLine(nameof(Generic) == nameof(Generic<object, object>)); // prints "True"


class Generic { public int field; }     //No generic arguments
class Generic<T> { public T field; }    //One generic argument
class Generic<T, U> { public U field; } //Two generic arguments
```
As is shown from the program's output, the type arguments of the type provided to the `nameof` operator have no effect on the resulting string.

However, writing something like...
```cs
Console.WriteLine(nameof(Generic<,>.field));
```
...is invalid. This leads to the need to use dummy type parameters, which must satisfy the restraints of the generic class, in order to get the correct name.

Furthermore, the `nameof` operator dissalows the usage of generic parameters when getting the name of a method group, thus creating an inconsistency between generic class members and methods.

## Proposal
I propose that the `nameof` operator permit the user to specify types using syntax similar to the specification of a generic type definitions in the `typeof` operator. This would allow the user to change the type requirements of a generic class without affecting any `nameof` operators which reference said class. Furthermore, refactoring of a generic class which shares its name with other generic classes with a different number of type arguments would still be able to account for `nameof` expressions, since the number of type arguments is supplied.
```cs
//Matches any class named "MyGeneric" which takes one type parameter
public const string MyGenericName = nameof(MyGeneric<>); 
//Matches any class named "My Generic" which takes two type parameters
public const string MyGenericName2 = nameof(MyGeneric<,>);
//Any field accessible with MyGeneric<ValidArgument> can be accessed
public const string MyGenericItem = nameof(MyGeneric<>.item);
```

**Comment by koszeggy** [comment]
I'm quite sure I've seen an issue for this in the dotnet/runtime repo but I cannot find it.

Btw. that's why I use a dummy internal struct named `_` so I can use `nameof(MyGeneric<_,_>)` to make it obvious that type arguments are not relevant.

**Comment by yaakov-h** [comment, marked as answer]
Duplicate of #702.

**Reply by koszeggy** [reply]
Ah, ok, so it was in the dotnet/roslyn repo then. I knew I saw it somewhere. :)

**Reply by jnm2** [reply]
Originally, but it's been in csharplang for four years after that.
